### PR TITLE
Minimal Lexing Support for Go Language

### PR DESCRIPTION
**Pull request by lkishalmi**
This is just provides simple syntax highlight for Go files, with Editor preferences.

I may able to add parser based error hints, which can be useful, maybe folding and navigatior support could be theoretically easy to be added, however not necessary for NB17.

### CODE REVIEW

**Comment by lkishalmi** [comment]
image

**Comment by lkishalmi** [comment]
Technically, it wouldn't be too hard to add navigator support to this. Later on if someone interested, could implement getting proper information from the parser. Even a full language support could be achievable. Unfortunately, I feel disgust when I see a Go code.

**Comment by lkishalmi** [comment]
Well, this one is just a showcase of my two new modules `lexer.antlr4` and `go.lang` for NB17.

AFAIK, Textmate based Syntax-highlight, provides:
- File icon/type association
- Syntax highlight

This one provides:
- File icon/type association
- Syntax highlight
- Custom Editor Preferences for Go File: Highlighting, Formatting
- Minimal Parser based error highlight.
- Code Folding

I never tried to use a Go LSP, I'm not really interested in that or in the Go language. I just need Go for my upcoming Terraform works,but for that the parser is enough.

**Comment by Chris2011** [comment]
I'm fine with this. I also think that ANTLR has more powerful options, problematic, what I think is kind of complex for beginners. At the end, mine is just a collection of those languages, @lkishalmi is direct inside of the core so we have a bit more options afaik. I would also say that we don't need more than two optiosn to add a new language. Antlr and textmate + lsp is fine if we have good resources for new comers to show them how to implement syntax highlighting, code completion (also context based), hints, formatting, code folding, etc. My two cents.

File: nbbuild/rat-exclusions.txt [hunk 1]
```diff
@@ -220,6 +220,7 @@ javafx/javafx2.project/src/org/netbeans/modules/javafx2/project/templates/**
 java/jshell.support/src/org/netbeans/modules/jshell/resources/consoleExample.jsh
 java/junit.ui/src/org/netbeans/modules/junit/ui/resources/*.template
 ide/languages.diff/src/org/netbeans/modules/languages/diff/DiffExample.diff
+ide/languages.go/src/org/netbeans/modules/languages/go/*.go
```
Comment (matthiasblaesing) [inline nbbuild/rat-exclusions.txt:223]: Please use the `licenseinfo.xml` to declare `GoExample.go` as `Apache-2.0-ASF`.

File: nbbuild/licenses/CC-BY-4.0 [hunk 2]
```diff
@@ -0,0 +1,429 @@
+Attribution-ShareAlike 4.0 International
```
Comment (matthiasblaesing) [inline nbbuild/licenses/CC-BY-4.0:1]: This is `CC-BY-SA-4.0`. Given the wording of the license in the attribution clause, this needs an entry in the notice section.

File: ide/languages.go/licenseinfo.xml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+-->
+<licenseinfo>
+    <fileset>
+        <file>src/org/netbeans/modules/languages/go/resources/go-blue.png</file>
+        <license ref="CC-BY-4.0" />
```
Comment (matthiasblaesing) [inline ide/languages.go/licenseinfo.xml:25]: As written below I find this problematic. This is missing the required attribution from CC-BY-SA-4.0.

I assume the logo was taken from the go lang repository. Wikimedia Commons holds a PD licensed copy: https://commons.wikimedia.org/wiki/File:Go_Logo_Blue.svg

If that is not acceptable, I'm willing to open my graphic programm and enter the two letters and choose the "right" color.

File: ide/languages.go/nbproject/project.properties [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+javac.source=1.8
+
+# As of 2nd Jan 2023, we use the Grammar from:
+# https://github.com/antlr/grammars-v4/tree/4794c0c5371ce6b0711a306256f8f7706394ec6d/golang
+#
+golang.grammar.version=4794c0c5371ce6b0711a306256f8f7706394ec6d
```
Comment (matthiasblaesing) [inline ide/languages.go/nbproject/project.properties:22]: This is not needed or is it? But it reminds me, that parsers have the tendency to create incompatible code when the grammar is modified. So either `go.lang` should be a friend-only export or we must be willing to drop compatibility in the future (i.e. raise major version).

**Review by matthiasblaesing — COMMENTED** [review]
I left a few inline comments, that I think need addressing before merging.

Go seems small enough, that it could be a nice example how to implement languages. Maybe I look at it at a later time.

File: ide/languages.go/licenseinfo.xml (hunk 3 above)
Comment (lkishalmi) [reply ide/languages.go/licenseinfo.xml:25]: Well it's a scaled down version the one in the bundle at: https://blog.golang.org/go-brand

The site says:
> Like this blog and all our gopher images, the slide themes are Creative Commons Attribution 4.0 licensed. 

The zip contains a `copyright.txt` with the content:
> All content within this zip file is Copyright 2018 The Go Authors. All rights reserved.

**Comment by lkishalmi** [comment]
Well, the workweek has started for me. @matthiasblaesing pointed out valid things on the Licensing part. Thanks again for your wisdom!

I moved the milestone to NB18, to give less headache to @neilcsmith-net when he cuts the release branch for NB17.

File: ide/languages.go/licenseinfo.xml (hunk 3 above)
Comment (lkishalmi) [reply ide/languages.go/licenseinfo.xml:25]: @matthiasblaesing I've just re-read the license and the ASF 3rd Party License policy. It seems the best would be to create an icon of our own.

The main issue I see here, is that this image is modified (resized) work, in which case we should abide to the share-alike clause, which we can't do in ASF. Unmodified work could be included. I've tried to use the original SVG, but NetBeans does not like that one.

Unfortunately, I'm really clumsy with graphics stuff, other than resizing. The TOML file icon is about my limit. So I'd be happy if you could type that 2 letters. (A 16x16 PNG is needed with transparent background) I think the selected blue would fit both light and dark LaF-s.

Comment (matthiasblaesing) [reply ide/languages.go/licenseinfo.xml:25]: Done: https://github.com/matthiasblaesing/netbeans/commit/437391c84b3178ee5533488e1a98c56a88633224

I documented how the "logo" was created in the SVG.

I noticed a warning when starting (registration of GoLanguage collides with antlr3 grammar one). This is my fix:

https://github.com/matthiasblaesing/netbeans/commit/50955a9a8a71dfac214be650ae69b5adda147f2c

Feel free to pull changes from https://github.com/matthiasblaesing/netbeans/commits/pr-5297 and squash them into your work.

**Comment by mbien** [comment]
created a https://github.com/apache/netbeans/labels/Go label. Since something tells me this won't be the last Go PR :)

File: ide/languages.go/nbproject/project.properties (hunk 4 above)
Comment (lkishalmi) [reply ide/languages.go/nbproject/project.properties]: I know it's a bit controversial topic nowadays, but I'd rather raise the release version on the go.lang module if needed.

Though I removed this stuff from here.

File: ide/languages.go/licenseinfo.xml (hunk 3 above)
Comment (lkishalmi) [reply ide/languages.go/licenseinfo.xml]: @matthiasblaesing Thanks for the commits. I really appreciate them!

File: nbbuild/rat-exclusions.txt (hunk 1 above)
Comment (lkishalmi) [reply nbbuild/rat-exclusions.txt]: Done.

File: nbbuild/licenses/CC-BY-4.0 (hunk 2 above)
Comment (lkishalmi) [reply nbbuild/licenses/CC-BY-4.0]: At the end I've removed the CC-BY-SA-4.0

**Review by matthiasblaesing — APPROVED** [review]
Looks good to me. I would do some squashing though, as IMHO the potentially problematic icon file should not become part of history. Feel free to just fold my changes into yours, its ok.

**Pull request merged**